src/Serilog.Extensions.Logging/Extensions/Logging/EventIdPropertyCache.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Serilog.Extensions.Logging
+{
+    using Microsoft.Extensions.Logging;
+    using Serilog.Events;
+
+    internal sealed class EventIdPropertyCache
+    {
+        private readonly object _createLock = new();
+        private readonly int _maxCapacity;
+        private readonly Dictionary<EventKey, LogEventProperty> _propertyCache;
+
+        private int count;
+
+        public EventIdPropertyCache(int maxCapacity)
+        {
+            _maxCapacity = maxCapacity;
+            _propertyCache = new Dictionary<EventKey, LogEventProperty>(capacity: maxCapacity);
+        }
+
+        public LogEventProperty GetOrCreateProperty(in EventId eventId)
+        {
+            var eventKey = new EventKey(eventId);
+
+            if (_propertyCache.TryGetValue(eventKey, out var cachedProperty))
+            {
+                return cachedProperty;
+            }
+
+            lock (_createLock)
+            {
+                // Double check under lock
+                if (_propertyCache.TryGetValue(eventKey, out cachedProperty))
+                {
+                    return cachedProperty;
+                }
+
+                cachedProperty = CreateCore(in eventKey);
+
+                if (count < _maxCapacity)
```

Comment:
Applications log a lot of things during start-up that they never log again. Perhaps clearing the cache when it reaches the size limit, and then (lazily) repopulating it, is worth considering, here? Bumping the cache size up to e.g. 256 items would help to avoid degenerate cases.

Reply:
I think we could consider making this cache shared (as a static class) among all loggers. Since `ConcurrentDictionary` is lock-free on reads, I believe this approach could perform as well as a per-logger cache.

> Perhaps clearing the cache when it reaches the size limit

This problem can be addressed by choosing the correct max-cached-items parameter. Speculatively, the total set of application and library event IDs can typically fit into a single cache. A maximum of 1024 items seems reasonable to me.

Implementing an LRU cache could help save some memory. However, this approach might introduce more lock contention, potentially impacting performance.

So, I'd like to suggest using the approach the Microsoft team implemented for caching in [FormattedLogValues](https://github.com/dotnet/runtime/blob/v8.0.10/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/FormattedLogValues.cs).

Reply:
Almost everything exactly the same as in FormattedLogValues source code.

```csharp
private const int MaxCachedProperties = 1024;

private static readonly ConcurrentDictionary<EventKey, LogEventProperty> s_propertyCache = new ();
private static int s_count;

public static LogEventProperty GetOrCreateProperty(in EventId eventId)
{
    var eventKey = new EventKey(eventId);

    LogEventProperty? property;

    if (s_count >= MaxCachedProperties)
    {
        if (!s_propertyCache.TryGetValue(eventKey, out property))
        {
            property = CreateCore(in eventKey);
        }
    }
    else
    {
        property = s_propertyCache.GetOrAdd(
            eventKey,
            static key =>
            {
                Interlocked.Increment(ref s_count);

                return CreateCore(in key);
            });
    }

    return property;
}
```

Actually, the factory in `GetOrAdd` might be called multiple times. I don't think this is a problem, as total correctness is not required in this case.